### PR TITLE
Feature/add authentication to check images on registries

### PR DESCRIPTION
**Pull request by dbalucas**
During the first test we have faced issues with docker registries where you have to provide authentication for.

Therefore different features were added:
- feature to authenticate against docker hub because it behaves special
- a logic for most know registries like github.io, quay.io, ...
- special handling for registries which require authentication, but credentials aren't available (usage of a default/general access-token)

This PR should add this logics and as well include test cases therefore.

### CODE REVIEW

File: .github/scripts/increment_version.py
Comment (raynigon) [inline .github/scripts/increment_version.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: setup.py
Comment (raynigon) [inline setup.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: src/lennybot/__init__.py
Comment (raynigon) [inline src/lennybot/__init__.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: src/lennybot/actions/download_resources.py
Comment (raynigon) [inline src/lennybot/actions/download_resources.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: src/lennybot/actions/update_dockerfile.py
Comment (raynigon) [inline src/lennybot/actions/update_dockerfile.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: src/lennybot/actions/update_image_tag.py
Comment (raynigon) [inline src/lennybot/actions/update_image_tag.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: src/lennybot/actions/update_yaml.py
Comment (raynigon) [inline src/lennybot/actions/update_yaml.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: src/lennybot/config/config.py
Comment (raynigon) [inline src/lennybot/config/config.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: src/lennybot/model/state.py
Comment (raynigon) [inline src/lennybot/model/state.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: src/lennybot/service/plan.py
Comment (raynigon) [inline src/lennybot/service/plan.py:1]: Out of scope of this PR.
For the next time please split this to a seperate PR.

File: version.txt
Comment (raynigon) [inline version.txt:1]: This should not be changed by a PR

**Review by raynigon — CHANGES_REQUESTED** [review]
Please split the encoding into a seperate PR and request review again

File: src/lennybot/check/docker_image_available.py [hunk 1]
```diff
@@ -66,20 +77,61 @@ def _parse_image(self):
...
         return DockerImage(match.group(4), match.group(5) + "/" + match.group(6), image_tag)
 
+    def _authenticate_on_registry(self, registry: str, realm: str, service: str, scope: str) -> str:
+        if registry not in self._container_config.registries.keys():
+            raise Exception(f"No credentials found for registry: {registry}")
+
+        registry_data = self._container_config.registries[registry]
+
+        if registry_data.password is None or "":
+            url = f"{realm}?scope={scope}&grant_type=password&service={service}&username={registry_data.username}&password={registry_data.password}&client_id=lennybot&access_type=offline"
```
Comment (raynigon) [inline src/lennybot/check/docker_image_available.py:87]:
```suggestion
            params = {"scope": scope, "grant_type": "password", "service": service, "username": registry_data.username, "password": registry_data.password, "client_id": "lennybot", "access_type": "offline"}
            url = f"{realm}?{url_encode_params(params)}"
```

File: src/lennybot/check/docker_image_available.py [hunk 2]
```diff
@@ -66,20 +77,61 @@ def _parse_image(self):
...
+            response = requests.get(url)
+        else:
+            url = f"{realm}?scope={scope}&grant_type=password&service={service}&client_id=lennybot&access_type=offline"
```
Comment (raynigon) [inline src/lennybot/check/docker_image_available.py:90]:
```suggestion
            params = {"scope": scope, "grant_type": "password", "service": service, "client_id": "lennybot", "access_type": "offline"}
            url = f"{realm}?{url_encode_params(params)}"
```

File: src/lennybot/check/docker_image_available.py [hunk 3]
```diff
@@ -66,20 +77,61 @@ def _parse_image(self):
...
+
+        if response.status_code == 401:
+            raise Exception("Error occured: Unauthenticated: ", response.status_code)
+
+        if response.status_code == 200:
+            token_data = response.json()
+            access_token = token_data.get("token")
+            return access_token
+
+        ## TODO: test for unauthenticated
```
Comment (raynigon) [inline src/lennybot/check/docker_image_available.py:101]: Test Case is missing

File: src/lennybot/check/docker_image_available.py [hunk 4]
```diff
@@ -66,20 +77,61 @@ def _parse_image(self):
             return DockerImage(None, match.group(2) + "/" + match.group(3), image_tag)
         return DockerImage(match.group(4), match.group(5) + "/" + match.group(6), image_tag)
 
+    def _authenticate_on_registry(self, registry: str, realm: str, service: str, scope: str) -> str:
+        if registry not in self._container_config.registries.keys():
+            raise Exception(f"No credentials found for registry: {registry}")
+
+        registry_data = self._container_config.registries[registry]
+
+        if registry_data.password is None or "":
```
Comment (raynigon) [inline src/lennybot/check/docker_image_available.py:86]: The handling of registries without credentials should not require a registry defined in the configuration file.

**Review by raynigon — CHANGES_REQUESTED** [review]
Please add test case and correct the logic for registries without credentials

File: src/lennybot/check/docker_image_available.py [hunk 5]
```diff
@@ -4,10 +4,11 @@
 
 import requests
 
-from ..config.config import LennyBotCheckConfig
+from ..config.config import LennyBotCheckConfig, LennyBotConfigContainerConfig, LennyBotConfigContainerRegistry
 from .icheck import ICheck
 
 PATTERN = r"(?:([\-\_\.\w]+)$)|(?:([\-\_\.\w]+)/([\-\_\.\w]+)$)|(?:([\-\.A-z0-9]+)/([\-\_\.\w]+)/([\-\_\.\w]+)$)"
+BLOCKED_REGISTRIES = []  # "docker.elastic.co"]  # Requires authentication, but is a public registry
```
Comment (raynigon) [inline src/lennybot/check/docker_image_available.py:11]: Still needed?

File: src/lennybot/check/docker_image_available.py [hunk 6]
```diff
@@ -4,10 +4,11 @@
...
```
Comment (raynigon) [inline src/lennybot/check/docker_image_available.py:12]: Create the `url_encode_params` function which takes a dictonary and returns an url encoded string of the parameters.

File: src/lennybot/check/docker_image_available.py [hunk 7]
```diff
@@ -66,20 +77,61 @@ def _parse_image(self):
...
+            res = requests.get(request_url, headers=headers)
+        else:
+            res = requests.get(request_url)
+
+        if res.status_code == 401 and access_token is None:
+            registry = image._registry
+            authenticate = res.headers["Www-Authenticate"]
+            realm = authenticate.split('realm="')[1].split('"')[0]
+            service = authenticate.split('service="')[1].split('"')[0]
+            scope = authenticate.split('scope="')[1].split('"')[0]
```
Comment (raynigon) [inline src/lennybot/check/docker_image_available.py:128]: Create Authentication Class and parsing function which returns an object of this class

File: test/lennybot.yaml [hunk 8]
```diff
@@ -8,6 +8,9 @@ github:
 container:
   registries:
     ghcr.io:
+      username: "USERNAME"
+      password: "<REDACTED>"
+    bracrebike.arc.io:
```
Comment (dbalucas) [inline test/lennybot.yaml:13]:
```suggestion

```

File: test/lennybot.yaml [hunk 9]
```diff
@@ -8,6 +8,9 @@ github:
...
       username: "<REDACTED>"
```
Comment (dbalucas) [inline test/lennybot.yaml:14]:
```suggestion
      username: "USERNAME"
```

**Review by dbalucas — COMMENTED** [review]
remove from config

File: src/lennybot/check/docker_image_available.py (hunk 3 above)
Comment (raynigon) [reply src/lennybot/check/docker_image_available.py:122]: @dbalucas please remove comment from here

File: test/config/test_config.py [hunk 10]
```diff
@@ -6,7 +6,9 @@
...
     def test_XXX(self):
-        os.environ["LB_CONTAINER_REGISTRY_ghcr.io_USERNAME"] = "lucas"
+        os.environ["LB_CONTAINER_REGISTRY_ghcr.io_USERNAME"] = "USERNAME"
         config = LennyBotConfig("test/lennybot.yaml")
-        self.assertEqual(1, len(config._container._registries))
-        self.assertEqual("lucas", config._container.registries["ghcr.io"].username)
+        # self.assertEqual(1, len(config._container._registries))
+        self.assertEqual("USERNAME", config._container.registries["ghcr.io"].username)
+
+    # pass
```
Comment (raynigon) [inline test/config/test_config.py:14]: @dbalucas please remove comment

File: version.txt
Comment (raynigon) [inline version.txt:1]: Please remove this file from the PR

File: versions.state.yaml
Comment (raynigon) [inline versions.state.yaml:1]: Please remove this file from the PR

File: src/lennybot/check/docker_image_available.py [hunk 11]
```diff
@@ -66,20 +93,72 @@ def _parse_image(self):
...
+        }
+
+        if registry in self._container_config.registries.keys():
+            registry_data = self._container_config.registries[registry]
+            params["password"] = registry_data.password
+            params["username"] = registry_data.username
+
+        url = f"{authentication_header.realm}?{urlencode(params)}"
+
+        response = requests.get(url)
```
Comment (raynigon) [inline src/lennybot/check/docker_image_available.py:112]:
```suggestion
        url = f"{authentication_header.realm}?{urlencode(params)}"
        response = requests.get(url)
```

**Review by raynigon — APPROVED** [review]
LGTM

**Pull request merged**